### PR TITLE
add energy consumption calculation

### PR DESCRIPTION
**Pull request by mirpedrol**
Example output trace file

```
task_id	energy_consumption	CO2e
1	20.328770833333333mWh	9.656166145833334mg
2	20.328770833333333mWh	9.656166145833334mg
3	20.328770833333333mWh	9.656166145833334mg
4	20.328770833333333mWh	9.656166145833334mg

```

Example summary trace file

```
The total CO2 emission is: 38.624664583333335mg
The total energy consumption is: 81.31508333333333uWh
```

### CODE REVIEW

File: plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy [hunk 1]
```diff
@@ -114,8 +115,23 @@ class CO2FootprintFactory implements TraceObserverFactory {
     }
 
 
+    private String convertToReadableUnits(double value, int unitIndex=4) {
+        def units = ['p', 'n', 'u', 'm', ' ', 'K', 'M', 'G', 'T', 'P', 'E']  // Units: pico, nano, micro, mili, 0, Kilo, Mega, Giga, Tera, Peta, Exa
```
Comment (skrakau) [inline plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy:119]: like this, if the unit is `gram` no unit is given out. Shouldn't `g` be given out as well?

**Review by skrakau — COMMENTED** [review]
Hi, this looks already great! I commented on two minor things to add.

File: plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy (hunk 1 above)
Comment (skrakau) [reply plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy:119]: ah, my mistake :D it's just the prefix

File: plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy [hunk 2]
```diff
@@ -114,8 +115,23 @@ class CO2FootprintFactory implements TraceObserverFactory {
     }
 
 
+    private String convertToReadableUnits(double value, int unitIndex=4) {
```
[the remark this answers is not in the document]
Comment (mirpedrol) [reply plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy:118]: Thanks for the suggestion! I will try to move this function somewhere where we can use it for different things, I found that I also need it in the CO2Record class for the record plots. 
I will also add it for total values 👍

File: plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy (hunk 1 above)
Comment (mirpedrol) [reply plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy:119]: This function outputs only the "prefix", the actual unit should be added afterwards, that way we can reuse it for different units, such as g, Wh...

Comment (mirpedrol) [reply plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy:119]: exactly, I didn't see your second message before replying 😄

File: plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy [hunk 3]
```diff
@@ -326,13 +348,19 @@ class CO2FootprintFactory implements TraceObserverFactory {
...
+
+
+
+            co2eRecords[taskId] = new CO2Record((Double) eConsumption, (Double) co2, trace.get('name').toString())
+            total_energy += eConsumption
             total_co2 += co2
 
             // save to the file
-            writer.send { PrintWriter it -> it.println("${taskId}\t${co2}"); it.flush() }
+            writer.send { PrintWriter it -> it.println("${taskId}\t${convertToReadableUnits(eConsumption,5)}Wh\t${convertToReadableUnits(co2)}g"); it.flush() }
```
[the remark this answers is not in the document]
Comment (mirpedrol) [reply plugins/nf-co2footprint/src/main/nextflow/co2footprint/CO2FootprintFactory.groovy:363]: I think it's good to let the user decide which fields to output :) but maybe out of the scope of this PR

**Comment by skrakau** [comment]
Great, LGTM!

**Review by skrakau — APPROVED** [review]

**Pull request merged**